### PR TITLE
Add compatibility with Ruby 3.4.1

### PR DESCRIPTION
**Pull request by herwinw**
I'm hesitant whether we should merge this, it might get in the way of some new features later on.

### CODE REVIEW

**Review by seven1m — DISMISSED** [review]

**Comment by seven1m** [comment]
I'll defer to you as our compatibility expert 😂

**Comment by herwinw** [comment]
I'm going to put this on hold for now. Most issues are related to frozen string literals, I'm well on my way of fixing those.

**Comment by herwinw** [comment]
This is getting along nicely. A few remarks:
* Docker is running on 3.4.0-rc1, since the stable images are not released yet (they're always a few days late due to Christmas), once they're released we can simply drop the last commit
* Macos is not yet running, there is no ruby3.4 in brew yet
* The new formatting of `Hash#inspect` (extra space on both sides of `=>`) is now different from the output of `ENV.inspect`, so this required a new function instead of delegating it to the Hash object. I made an upstream report, since this looks weird and inconsistent (https://bugs.ruby-lang.org/issues/20982).
* The statement `defined?({1=>@baz})` where the ivar does not yet exist now differs from MRI, the Natalie tests have an ugly workaround to be able to run this test with MRI 3.3, MRI 3.4 and Natalie.

**Comment by herwinw** [comment]
It's passing all tests now, so it's finally ready to merge.

File: .github/workflows/test.yml
Comment (seven1m) [inline .github/workflows/test.yml:86]: Hmmm, for a short period yesterday, the `/opt/homebrew/opt/ruby` path didn't exist. But I guess they fixed it. 🤷‍♂️

Comment (herwinw) [reply .github/workflows/test.yml:86]: The same path (without the explicit version) is used in an earlier debug step to print the ruby version, so this makes it more consistent as well.

File: test/natalie/enumerator_test.rb
Comment (seven1m) [inline test/natalie/enumerator_test.rb:196]: There appears to be no difference between these two blocks -- the two ruby versions behave the same?

**Review by seven1m — APPROVED** [review]
That was a huge effort! Well done.

So. Many. Backticks. 😆

File: test/natalie/enumerator_test.rb
Comment (herwinw) [reply test/natalie/enumerator_test.rb:196]: No, this one resulted in an error when using MRI 3.3 as a host version. Just one more backtick

**Pull request merged**
